%YAML 1.2
# SPDX-FileCopyrightText: NVIDIA CORPORATION & AFFILIATES
# Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0
---
name: sync
components:
- name: camera_message_rx
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_message_rx
    min_sum: 1
- name: camera_info_message_rx
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 2
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: camera_info_message_rx
    min_sum: 1
- name: camera_message_tx
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_message_tx
    min_size: 1
- name: camera_info_message_tx
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_info_message_tx
    min_size: 1
- type: nvidia::isaac::CameraInfoSynchronization
  parameters:
    camera_message_rx: camera_message_rx
    camera_info_message_rx: camera_info_message_rx
    camera_message_tx: camera_message_tx
    camera_info_message_tx: camera_info_message_tx
    use_latest_camera_info: false
    drop_old_messages: false
---
name: input_compositor
components:
- name: image_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: image_in
    min_size: 1
- name: cam_info_in
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: cam_info_in
    min_size: 1
- name: out
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: out
    min_size: 1
- name: compositor
  type: nvidia::isaac_ros::MessageCompositor
  parameters:
    image_rx: image_in
    camera_model_rx: cam_info_in
    tx: out
---
name: imageResizer
components:
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- name: data_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: data_transmitter
    min_size: 1
- name: pool
  type: nvidia::gxf::BlockMemoryPool
  parameters:
    storage_type: 1
    block_size: 7372800
    num_blocks: 40
- name: cuda_stream
  type: nvidia::gxf::CudaStreamPool
  parameters:
    stream_flags: 1
    stream_priority: 0
    reserved_size: 1
    max_size: 1
    nvtx_identifier: "nitros_resize_node_cuda_stream"
- name: StreamResize
  type: nvidia::isaac::tensor_ops::Resize
  parameters:
    output_width: 400
    output_height: 300
    interp_type: "linear"
    border_type: "zero"
    keep_aspect_ratio: false
    receiver: data_receiver
    transmitter: data_transmitter
    pool: pool
    input_adapter: resource/adapter
    output_adapter: resource/adapter
    output_name: "image"
    stream_pool: cuda_stream
- type: nvidia::gxf::MemoryAvailableSchedulingTerm
  parameters:
    allocator: pool
    min_blocks: 1
---
name: streamsync
components:
- name: rx
  type: nvidia::gxf::DoubleBufferReceiver
- name: tx
  type: nvidia::gxf::DoubleBufferTransmitter
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: rx
    min_size: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: tx
    min_size: 1
- type: nvidia::gxf::CudaStreamSync
  parameters:
    rx: rx
    tx: tx
---
name: splitter
components:
- name: data_receiver
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: data_receiver
    min_size: 1
- name: image_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: image_transmitter
    min_size: 1
- name: camera_model_transmitter
  type: nvidia::gxf::DoubleBufferTransmitter
  parameters:
    capacity: 1
- type: nvidia::gxf::DownstreamReceptiveSchedulingTerm
  parameters:
    transmitter: camera_model_transmitter
    min_size: 1
- name: splitter
  type: nvidia::isaac_ros::MessageSplitter
  parameters:
    rx: data_receiver
    image_tx: image_transmitter
    camera_model_tx: camera_model_transmitter
---
name: image_sink
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: signal
---
name: camera_info_sink
components:
- name: signal
  type: nvidia::gxf::DoubleBufferReceiver
  parameters:
    capacity: 1
- type: nvidia::gxf::MessageAvailableSchedulingTerm
  parameters:
    receiver: signal
    min_size: 1
- name: sink
  type: nvidia::isaac_ros::MessageRelay
  parameters:
    source: signal
---
name: resource
components:
- name: adapter
  type: nvidia::isaac::tensor_ops::ImageAdapter
  parameters:
    message_type: "VideoBuffer"
- name: stream
  type: nvidia::isaac::tensor_ops::TensorStream
  parameters:
    backend_type: "VPI"
    engine_type: "GPU"
---
components:
- type: nvidia::gxf::Connection
  parameters:
    source: sync/camera_message_tx
    target: input_compositor/image_in
- type: nvidia::gxf::Connection
  parameters:
    source: sync/camera_info_message_tx
    target: input_compositor/cam_info_in
- type: nvidia::gxf::Connection
  parameters:
    source: input_compositor/out
    target: imageResizer/data_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: imageResizer/data_transmitter
    target: streamsync/rx
- type: nvidia::gxf::Connection
  parameters:
    source: streamsync/tx
    target: splitter/data_receiver
- type: nvidia::gxf::Connection
  parameters:
    source: splitter/image_transmitter
    target: image_sink/signal
- type: nvidia::gxf::Connection
  parameters:
    source: splitter/camera_model_transmitter
    target: camera_info_sink/signal
---
name: utils
components:
- name: clock
  type: nvidia::gxf::RealtimeClock
- type: nvidia::gxf::EventBasedScheduler
  parameters:
    clock: clock
    stop_on_deadlock: false
    worker_thread_number: 2
    worker_thread_name_id: "nitros_resize_node"
- type: nvidia::gxf::JobStatistics
  parameters:
    clock: clock
